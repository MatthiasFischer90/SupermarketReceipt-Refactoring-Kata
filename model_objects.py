from enum import Enum
from typing import Optional


class ProductUnit(Enum):
    EACH = 1
    KILO = 2


class Product:
    def __init__(self, name: str, unit: ProductUnit):
        self.name = name
        self.unit = unit


class SpecialOfferType(Enum):
    THREE_FOR_TWO = 1
    PERCENT_DISCOUNT = 2
    TWO_FOR_AMOUNT = 3
    FIVE_FOR_AMOUNT = 4


class Offer:
    def __init__(
        self,
        offer_type: SpecialOfferType,
        product: Product,
        optional_argument: Optional[float],
    ):
        self.offer_type = offer_type
        self.product = product
        self.optional_argument = optional_argument


class Discount:
    def __init__(self, product: Product, description: str, discount_amount_cents: int):
        self.product = product
        self.description = description
        self.discount_amount_counts = discount_amount_cents
