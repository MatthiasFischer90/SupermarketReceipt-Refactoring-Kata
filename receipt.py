from typing import Union
from model_objects import Discount, Product


class ReceiptItem:
    def __init__(
        self,
        product: Product,
        quantity: Union[int, float],
        price_cents: int,
        total_price_cents: int,
    ):
        self.product = product
        self.quantity = quantity
        self.price_cents = price_cents
        self.total_price_cents = total_price_cents


class Receipt:
    def __init__(self):
        self._items: list[ReceiptItem] = []
        self._discounts: list[Discount] = []

    def get_total_price_cents(self) -> int:
        total = 0
        for item in self.items:
            total += item.total_price_cents
        for discount in self.discounts:
            total += discount.discount_amount_cents
        return total

    def add_product(
        self,
        product: Product,
        quantity: Union[int, float],
        price_cents: int,
        total_price_cents: int,
    ) -> None:
        self._items.append(
            ReceiptItem(
                product=product,
                quantity=quantity,
                price_cents=price_cents,
                total_price_cents=total_price_cents,
            )
        )

    def add_discounts(self, discounts: list[Discount]) -> None:
        self._discounts += discounts

    @property
    def items(self):
        return self._items[:]

    @property
    def discounts(self):
        return self._discounts[:]
