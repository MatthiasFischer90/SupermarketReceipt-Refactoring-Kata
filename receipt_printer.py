"""Module that contains the classes responsible for creating string representations of Receipts."""

from abc import ABC, abstractmethod

from model_objects import Discount, ProductUnit
from receipt import Receipt, ReceiptItem


class ReceiptPrinter(ABC):
    @abstractmethod
    def print_receipt(self, receipt: Receipt) -> str:
        """Creates a text representation of a given Receipt.

        Args:
            receipt (Receipt): The Receipt for which to create the text representation.

        Returns:
            str: The text representation of the given Receipt.
        """
        pass


class TextReceiptPrinter(ReceiptPrinter):
    """Class used to create text file representation of given Receipt.

    The resulting text representation is formatted similar to how the
    text would be formatted on a real-life receipt slip.
    """

    def __init__(self, columns: int = 40):
        if columns < 1:
            raise ValueError(f"columns must be positive integer, but got {columns}!")
        self.columns = columns

    def print_receipt(self, receipt: Receipt) -> str:
        result = ""
        for item in receipt.items:
            receipt_item = self._print_receipt_item(item=item)
            result += receipt_item

        for discount in receipt.discounts:
            discount_presentation = self._print_discount(discount=discount)
            result += discount_presentation

        result += "\n"
        result += self._present_total(receipt=receipt)
        return result

    def _print_receipt_item(self, item: ReceiptItem) -> str:
        total_price_printed = self._print_price(price_cents=item.total_price_cents)
        name = item.product.name
        line = self._format_line_with_whitespace(name=name, value=total_price_printed)
        if item.quantity != 1:
            line += f"  {self._print_price(price_cents=item.price_cents)} * {self._print_quantity(item=item)}\n"
        return line

    def _format_line_with_whitespace(self, name: str, value: str) -> str:
        line = name
        # leave at least one whitespace between name and value
        line += " "
        whitespace_size = self.columns - len(line) - len(value)
        for _ in range(whitespace_size):
            line += " "
        line += value
        line += "\n"
        return line

    def _print_price(self, price_cents: int) -> str:
        return "%.2f" % (price_cents / 100)

    def _print_quantity(self, item: ReceiptItem) -> str:
        if ProductUnit.EACH == item.product.unit:
            return str(item.quantity)
        else:
            return "%.3f" % item.quantity

    def _print_discount(self, discount: Discount) -> str:
        name = f"{discount.description} ({discount.product.name})"
        value = self._print_price(price_cents=discount.discount_amount_cents)
        return self._format_line_with_whitespace(name=name, value=value)

    def _present_total(self, receipt: Receipt) -> str:
        name = "Total: "
        value = self._print_price(price_cents=receipt.get_total_price_cents())
        return self._format_line_with_whitespace(name=name, value=value)


class HtmlReceiptPrinter(ReceiptPrinter):
    """Class used to create HTML representation of given Receipt."""

    HTML_PREFIX = """<!DOCTYPE html>
<html>
  <head>
    <title>Receipt</title>
    <style>
      table, td, th { border : 1px solid black; }
      table { margin-bottom: 10px; margin-top: 10px;}
      th { padding : 13px; }
      td { padding : 15px; }
    </style>
  </head>
  <body>
    """

    HTML_SUFFIX = """  </body>
</html>"""

    def print_receipt(self, receipt: Receipt) -> str:
        body_indentation = 4
        result = self.HTML_PREFIX
        result += self._print_item_table(receipt=receipt, indentation=body_indentation)
        result += self._print_discount_table(
            receipt=receipt, indentation=body_indentation
        )
        result += self._print_total(receipt=receipt, indentation=body_indentation)
        result += self.HTML_SUFFIX
        return result

    def _print_indentation(self, indentation: int) -> str:
        return " " * indentation

    def _get_price_string(self, price_cents: int) -> str:
        return "%.2f" % (price_cents / 100)

    def _get_quantity_string(self, item: ReceiptItem) -> str:
        if ProductUnit.EACH == item.product.unit:
            return str(item.quantity)
        else:
            return "%.3f" % item.quantity

    def _print_item_table(self, receipt: Receipt, indentation: int) -> str:
        if len(receipt.items) == 0:
            return ""

        content = "<table>\n"
        content += self._print_item_table_headers(indentation=(indentation + 2))
        for item in receipt.items:
            content += self._print_item_table_row(
                item=item, indentation=(indentation + 2)
            )
        content += self._print_indentation(indentation=indentation)
        content += "</table>\n"
        return content

    def _print_item_table_headers(self, indentation: int) -> str:
        result = self._print_indentation(indentation=indentation)
        result += "<tr>\n"
        result += self._print_indentation(indentation=(indentation + 2))
        result += "<th>Product name</th>\n"
        result += self._print_indentation(indentation=(indentation + 2))
        result += "<th>Unit price (EUR)</th>\n"
        result += self._print_indentation(indentation=(indentation + 2))
        result += "<th>Quantity</th>\n"
        result += self._print_indentation(indentation=(indentation + 2))
        result += "<th>Total price (EUR)</th>\n"
        result += self._print_indentation(indentation=indentation)
        result += "</tr>\n"
        return result

    def _print_item_table_row(self, item: ReceiptItem, indentation: int) -> str:
        row = self._print_indentation(indentation=indentation)
        row += "<tr>\n"
        row += self._print_product_name(item=item, indentation=(indentation + 2))
        row += self._print_item_unit_price(item=item, indentation=(indentation + 2))
        row += self._print_item_quantity(item=item, indentation=(indentation + 2))
        row += self._print_item_total_price(item=item, indentation=(indentation + 2))
        row += self._print_indentation(indentation=indentation)
        row += "</tr>\n"
        return row

    def _print_product_name(self, item: ReceiptItem, indentation: int) -> str:
        cell = self._print_indentation(indentation=indentation)
        cell += f"<td>{item.product.name}</td>\n"
        return cell

    def _print_item_unit_price(self, item: ReceiptItem, indentation: int) -> str:
        price_euros_string = self._get_price_string(price_cents=item.price_cents)
        cell = self._print_indentation(indentation=indentation)
        cell += f"<td>{price_euros_string}</td>\n"
        return cell

    def _print_item_quantity(self, item: ReceiptItem, indentation: int) -> str:
        quantity_string = self._get_quantity_string(item=item)
        cell = self._print_indentation(indentation=indentation)
        cell += f"<td>{quantity_string}</td>\n"
        return cell

    def _print_item_total_price(self, item: ReceiptItem, indentation: int) -> str:
        price_euros_string = self._get_price_string(price_cents=item.total_price_cents)
        cell = self._print_indentation(indentation=indentation)
        cell += f"<td>{price_euros_string}</td>\n"
        return cell

    def _print_discount_table(self, receipt: Receipt, indentation: int) -> str:
        if len(receipt.discounts) == 0:
            return ""

        content = self._print_indentation(indentation=indentation)
        content += "<table>\n"
        content += self._print_discount_table_headers(indentation=(indentation + 2))
        for discount in receipt.discounts:
            content += self._print_discount_table_row(
                discount=discount, indentation=(indentation + 2)
            )
        content += self._print_indentation(indentation=indentation)
        content += "</table>\n"
        return content

    def _print_discount_table_headers(self, indentation: int) -> str:
        row = self._print_indentation(indentation=indentation)
        row += "<tr>\n"
        row += self._print_indentation(indentation=(indentation + 2))
        row += "<th>Discount description</th>\n"
        row += self._print_indentation(indentation=(indentation + 2))
        row += "<th>Discount value (EUR)</th>\n"
        row += self._print_indentation(indentation=indentation)
        row += "</tr>\n"
        return row

    def _print_discount_table_row(self, discount: Discount, indentation: int) -> str:
        row = self._print_indentation(indentation=indentation)
        row += "<tr>\n"
        row += self._print_discount_description(
            discount=discount, indentation=(indentation + 2)
        )
        row += self._print_discount_value(
            discount=discount, indentation=(indentation + 2)
        )
        row += self._print_indentation(indentation=indentation)
        row += "</tr>\n"
        return row

    def _print_discount_description(self, discount: Discount, indentation: int) -> str:
        cell = self._print_indentation(indentation=indentation)
        cell += f"<td>{discount.description}</td>\n"
        return cell

    def _print_discount_value(self, discount: Discount, indentation: int) -> str:
        discount_value_euros_string = self._get_price_string(
            price_cents=discount.discount_amount_cents
        )
        cell = self._print_indentation(indentation=indentation)
        cell += f"<td>{discount_value_euros_string}</td>\n"
        return cell

    def _print_total(self, receipt: Receipt, indentation: int) -> str:
        total_price_euros_string = self._get_price_string(
            price_cents=receipt.get_total_price_cents()
        )
        line = self._print_indentation(indentation=indentation)
        line += f"<p>Total: {total_price_euros_string}</p>\n"
        return line
