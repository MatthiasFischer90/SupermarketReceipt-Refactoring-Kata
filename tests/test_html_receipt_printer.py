"""This module contains tests for the HtmlReceiptPrinter class and creates output files.

The tests in this module create files that are compared to the expected output files
in the approved_files subdirectory.
"""

from approvaltests import verify
from model_objects import Discount, Product, ProductUnit
from receipt import Receipt
from receipt_printer import HtmlReceiptPrinter


def test_quantity_one():
    receipt = Receipt()
    toothbrush = Product(name="toothbrush", unit=ProductUnit.EACH)
    receipt.add_product(
        product=toothbrush, quantity=1, price_cents=99, total_price_cents=99
    )
    verify(HtmlReceiptPrinter().print_receipt(receipt=receipt))


def test_quantity_two():
    receipt = Receipt()
    toothbrush = Product(name="toothbrush", unit=ProductUnit.EACH)
    receipt.add_product(
        product=toothbrush, quantity=2, price_cents=99, total_price_cents=198
    )
    verify(HtmlReceiptPrinter().print_receipt(receipt=receipt))


def test_loose_weight():
    receipt = Receipt()
    apples = Product(name="apples", unit=ProductUnit.KILO)
    receipt.add_product(
        product=apples,
        quantity=2.3,
        price_cents=199,
        total_price_cents=round(199 * 2.3),
    )
    verify(HtmlReceiptPrinter().print_receipt(receipt=receipt))


def test_total():
    receipt = Receipt()
    toothbrush = Product(name="toothbrush", unit=ProductUnit.EACH)
    apples = Product(name="apples", unit=ProductUnit.KILO)
    receipt.add_product(
        product=toothbrush, quantity=2, price_cents=99, total_price_cents=198
    )
    receipt.add_product(
        product=apples,
        quantity=0.75,
        price_cents=199,
        total_price_cents=round(199 * 0.75),
    )
    verify(HtmlReceiptPrinter().print_receipt(receipt=receipt))


def test_discounts():
    receipt = Receipt()
    apples = Product(name="apples", unit=ProductUnit.KILO)
    receipt.add_discounts(
        [Discount(product=apples, description="3 for 2", discount_amount_cents=-99)]
    )
    verify(HtmlReceiptPrinter().print_receipt(receipt=receipt))


def test_whole_receipt():
    receipt = Receipt()
    toothbrush = Product(name="toothbrush", unit=ProductUnit.EACH)
    apples = Product(name="apples", unit=ProductUnit.KILO)
    receipt.add_product(
        product=toothbrush, quantity=3, price_cents=99, total_price_cents=297
    )
    receipt.add_product(
        product=apples,
        quantity=0.75,
        price_cents=199,
        total_price_cents=round(199 * 0.75),
    )
    receipt.add_discounts(
        [Discount(product=toothbrush, description="3 for 2", discount_amount_cents=-99)]
    )
    verify(HtmlReceiptPrinter().print_receipt(receipt=receipt))
