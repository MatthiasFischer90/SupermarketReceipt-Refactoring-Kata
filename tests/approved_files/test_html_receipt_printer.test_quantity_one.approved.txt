<!DOCTYPE html>
<html>
  <head>
    <title>Receipt</title>
    <style>
      table, td, th { border : 1px solid black; }
      table { margin-bottom: 10px; margin-top: 10px;}
      th { padding : 13px; }
      td { padding : 15px; }
    </style>
  </head>
  <body>
    <table>
      <tr>
        <th>Product name</th>
        <th>Unit price (EUR)</th>
        <th>Quantity</th>
        <th>Total price (EUR)</th>
      </tr>
      <tr>
        <td>toothbrush</td>
        <td>0.99</td>
        <td>1</td>
        <td>0.99</td>
      </tr>
    </table>
    <p>Total: 0.99</p>
  </body>
</html>
